### DISCUSSION TITLE
GetAGLevel in CustomGroundCallback not being invoked

### DISCUSSION THREAD

**Discussion by CarloBlanch1967** [JSBSim-Team/jsbsim#1174] [General]
Problem Description: I'm implementing a custom CustomGroundCallback class in JSBSim to handle ground interactions for landing gear. However, the GetAGLevel function is never called during the simulation, even though I followed examples and the model configuration seems correct.

Here’s my implementation:

```c++
class CustomGroundCallback : public JSBSim::FGGroundCallback {
public:
    double terrain_elevation_nose;
    double terrain_elevation_rear;

    CustomGroundCallback(double nose, double rear)
        : terrain_elevation_nose(nose), terrain_elevation_rear(rear) {}

    double GetAGLevel(double t, const JSBSim::FGLocation& location,
                      JSBSim::FGLocation& contact,
                      JSBSim::FGColumnVector3& normal,
                      JSBSim::FGColumnVector3& v,
                      JSBSim::FGColumnVector3& w) const override {
        
        if (fabs(location(0) - 158.0) < 0.01) {
            return terrain_elevation_nose;
        } else {
            return terrain_elevation_rear;
        }
    }
};

```

Question: Is there any additional configuration needed to ensure JSBSim uses this callback for gear interactions?

Thank you in advance for your help!

**Comment by seanmcleod** [comment]
Okay, but you haven't shown your code where you set your custom ground callback as the callback to be used, e.g.

https://github.com/JSBSim-Team/jsbsim/blob/0f00b806a74027d71a3e758bbb1e177e4b6130f2/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L402

Otherwise the default ground callback supplied by JSBSim will be used:

https://github.com/JSBSim-Team/jsbsim/blob/0f00b806a74027d71a3e758bbb1e177e4b6130f2/src/models/FGInertial.cpp#L76

**Comment by CarloBlanch1967** [comment]
Thank You!  

Now It works in this way:
```c++
std::shared_ptr<JSBSim::FGInertial> Inertial = FDMExec->GetInertial();
Inertial->SetGroundCallback(new CustomGroundCallback(mesh_probe, mesh_probe_rear));
```
Now it works, but it doesn't work well. The point of contact is not established, the plane falls vertically as if there were no ground under the wheels. I logged the elevation I'm sending, and the numbers are correct.
It almost seems like the elevation data isn't being passed to the landing gear.
Is there something wrong with the class?
mesh_probe and mesh_probe_rear are two doubles that I import from the visual, and they are expressed in feet.

**Comment by CarloBlanch1967** [comment]
I logged location(0) and it is always at 0.00.

**Comment by CarloBlanch1967** [comment]
Should I subtract the elevation from the geodetic altitude before return?

[https://jsbsim-team.github.io/jsbsim/FGGroundCallback_8cpp_source.html](url)

```c++
double FGDefaultGroundCallback::GetAGLevel(double t, const FGLocation& loc,
                                     FGLocation& contact, FGColumnVector3& normal,
                                     FGColumnVector3& vel, FGColumnVector3& angularVel) const
 {
   vel.InitMatrix();
   angularVel.InitMatrix();
   FGLocation l = loc;
   l.SetEllipse(a,b);
   double latitude = l.GetGeodLatitudeRad();
   double cosLat = cos(latitude);
   double longitude = l.GetLongitude();
   normal = FGColumnVector3(cosLat*cos(longitude), cosLat*sin(longitude),
                            sin(latitude));
   contact.SetEllipse(a, b);
   contact.SetPositionGeodetic(longitude, latitude, mTerrainElevation);
   return l.GetGeodAltitude() - mTerrainElevation;
 }
```

**Comment by seanmcleod** [comment]
> plane falls vertically as if there were no ground under the wheels.

Do you have retractable gear, and if so, is the gear possibly retracted?

> Should I subtract the elevation from the geodetic altitude before return?

Remember, the function you're implementing needs to return the AGL between the `loc`, typically of the gear, some contact point defined for the model, or the cg, and then also return via the pass by reference parameters the location of what the contact point would be, the normal of the terrain at this contact point and whether the terrain at this contact point has any linear and/or angular velocity, typically only an issue for things like aircraft carriers.

**Reply by CarloBlanch1967** [reply]
> Do you have retractable gear, and if so, is the gear possibly retracted?

I don't think so because with this code

`Propagate->SetTerrainElevation(mesh_probe_rear + maxGearLenght);`

it works perfectly, it adheres to the mesh and follows it faithfully, but the aircraft doesn't tilt because there is only one contact point, I suppose.

**Comment by seanmcleod** [comment]
> I logged location(0) and it is always at 0.00.

Not sure exactly what you mean by `location(0)`.

Now when `const FGLocation& loc` is passed into your ground callback method, it is only initialised with ECEF coordinates, for example, this is what I see in the debugger.

```sh
loc
{mECLoc={data=0x000000000014eb68 {-7559866.5600183215, 12098308.575596180, 15298471.380769854} } mLon=...}
    JSBSim::FGJSBBase: {...}
    mECLoc: {data=0x000000000014eb68 {-7559866.5600183215, 12098308.575596180, 15298471.380769854} }
    mLon: 0.0000000000000000
    mLat: 0.0000000000000000
    mRadius: 0.0000000000000000
    mGeodLat: 0.0000000000000000
    GeodeticAltitude: 0.0000000000000000
    mTl2ec: {data=0x000000000014eba8 {0.0000000000000000, 0.0000000000000000, 0.0000000000000000, 0.0000000000000000, ...} }
    mTec2l: {data=0x000000000014ebf0 {0.0000000000000000, 0.0000000000000000, 0.0000000000000000, 0.0000000000000000, ...} }
    a: 1.0000000000000000
    e2: 0.0000000000000000
    c: 0.0000000000000000
    ec: 1.0000000000000000
    ec2: 1.0000000000000000
    mCacheValid: false
    mEllipseSet: false
```

You will notice the following in the default ground callback that you posted above, i.e. it makes a copy of `loc`, and then sets the ellipse parameters, so that it can then query this copy for lat, lon etc.

```c++
   FGLocation l = loc;
   l.SetEllipse(a,b);
   double latitude = l.GetGeodLatitudeRad();
```

So this is what I see in the debugger for `l`.

```sh
l
{mECLoc={data=0x000000000014e838 {-7559866.5600183215, 12098308.575596180, 15298471.380769854} } mLon=...}
    JSBSim::FGJSBBase: {...}
    mECLoc: {data=0x000000000014e838 {-7559866.5600183215, 12098308.575596180, 15298471.380769854} }
    mLon: 2.1293019443065497
    mLat: 0.82030435957987735
    mRadius: 20918027.616959054
    mGeodLat: 0.82364901223183440
    GeodeticAltitude: 29995.922120246123
    mTl2ec: {data=0x000000000014e878 {0.38876503631779880, -0.62215375568271758, 0.67954871115079740, -0.84804796003419902, ...} }
    mTec2l: {data=0x000000000014e8c0 {0.38876503631779880, -0.84804796003419902, 0.36010610105724961, -0.62215375568271758, ...} }
    a: 20925646.325460002
    e2: 0.0066943800085510485
    c: 140084.22842716915
    ec: 0.99664718932601670
    ec2: 0.99330561999144895
    mCacheValid: true
    mEllipseSet: true
```

**Comment by CarloBlanch1967** [comment]
I'm sorry, I can't do it, my skills are not sufficient, but thank you anyway.
